h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/task/XGBoostUploadMatrixTask.java

```diff
@@ -1,42 +1,373 @@
         this.remoteNodes = remoteNodes;
         this.https = https;
         this.contextPath = contextPath;
         this.userName = userName;
         this.password = password;
+
+        this.modelInfo = model.model_info();
+        this.parms = model._parms;
+        this.sparse = model._output._sparse;
+        this.train = train;
+    }
+    
+    private XGBoostHttpClient makeClient() {
+        String remoteUri = remoteNodes[H2O.SELF.index()] + contextPath;
+        return new XGBoostHttpClient(remoteUri, https, userName, password);
     }
 
     @Override
     protected void execute() {
-        XGBoostHttpClient client = new XGBoostHttpClient(remoteNodes[H2O.SELF.index()] + contextPath, https, userName, password);
-        LOG.info("Preparing matrix part to upload.");
-        MatrixLoader.DMatrixProvider data = matrixLoader.makeLocalMatrix();
-        client.uploadObject(_modelKey, "matrix", data);
+        XGBoostHttpClient client = makeClient();
+        LOG.info("Starting matrix upload for " + _modelKey);
+        long start = System.currentTimeMillis();
+        assert modelInfo.dataInfo() != null;
+        int[] chunks = VecUtils.getLocalChunkIds(train.anyVec());
+        final Vec responseVec = train.vec(parms._response_column);
+        final Vec weightVec = train.vec(parms._weights_column);
+        final Vec offsetsVec = train.vec(parms._offset_column);
+        final int[] nRowsByChunk = new int[chunks.length];
+        final long nRowsL = sumChunksLength(chunks, responseVec, weightVec, nRowsByChunk);
+        if (nRowsL > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("XGBoost currently doesn't support datasets with more than " +
+                Integer.MAX_VALUE + " per node. " +
+                "To train a XGBoost model on this dataset add more nodes to your H2O cluster and use distributed training.");
+        }
+        final int nRows = (int) nRowsL;
+        MatrixData matrixData = new MatrixData(nRows, weightVec, offsetsVec);
+        if (sparse) {
+            LOG.debug("Treating matrix as sparse.");
+            matrixData.shape = modelInfo.dataInfo().fullN();
+            matrixData.actualRows = csr(
+                client, chunks, weightVec, offsetsVec, responseVec, modelInfo.dataInfo(), 
+                matrixData.resp, matrixData.weights, matrixData.offsets
+            );
+        } else {
+            LOG.debug("Treating matrix as dense.");
+            matrixData.actualRows = dense(
+                client, chunks, nRows, nRowsByChunk, weightVec, offsetsVec, responseVec, modelInfo.dataInfo(),
+                matrixData.resp, matrixData.weights, matrixData.offsets
+            );
+        }
+        client.uploadObject(_modelKey, RemoteXGBoostUploadServlet.RequestType.matrixData, matrixData);
+        LOG.debug("Matrix upload finished in " + ((System.currentTimeMillis() - start) / 1000d));
+    }
+
+    public static class MatrixData implements Serializable {
```

Comment:
Since this holds data - Iced serialization is much more efficient, why are we using Serializables?